README.md
## rey
The Rust version of [hey](https://github.com/rakyll/hey), a tool for sending load to web application, like ab.

The purpose of this project is purely for learning and understanding the Rust language in a more practical way.
We use [reqwest](https://github.com/seanmonstar/reqwest) as the HTTP client. Due to the nature of reqwest, certain low-level HTTP statistics such as detailed timing data are not currently supported.

## Usage
```
Usage: rey [OPTIONS] <URL>

Arguments:
  <URL>  

Options:
  -n <REQUESTS>               Name of the person to greet [default: 200]
  -c <WORKERS>                Number of workers to run concurrently. Total number of requests cannot be smaller than the concurrency level [default: 50]
  -q <RATE LIMIT>             Rate limit, in queries per second (QPS) per worker
  -z <Duration>               Duration of application to send requests. When duration is reached, application stops and exits. If duration is specified, n is ignored. Examples: -z 10s -z 3m
  -m <METHOD>                 HTTP method, one of GET, POST, PUT, DELETE, HEAD, OPTIONS [default: GET]
  -H <HEADERS>                Custom HTTP header. You can specify as many as needed by repeating the flag. For example, -H "Accept: text/html" -H "Content-Type: application/xml"
  -t <TIMEOUT>                Timeout for each request in seconds. Use 0 for infinite [default: 20]
  -A <ACCEPT HEADER>          HTTP Accept header
  -T <CONTENT-TYPE>           Content-type, defaults to "text/html" [default: text/html]
  -U <USER AGENT>             User-Agent, defaults to version "rey/0.1.0"
  -d <BODY>                   HTTP request body
  -D <FILE>                   HTTP request body from file. For example, /home/user/file.txt or ./file.txt
  -a <USERNAME:PASSWORD>      Basic authentication, username:password
  -x <PROXY>                  HTTP Proxy address as host:port
      --host <HOST>           
      --disable-redirects     
  -h, --help                  Print help
  -V, --version               Print version
```
## Output
```
Summary:
  Total:  3.1956 secs
  Slowest:  0.4934 secs
  Fastest:  0.0084 secs
  Average:  0.2426 secs
  Requests/sec:  31.2928
  
  Total data:   500 B bytes
  Size/request: 5 B bytes

Response time histogram:
0.008 [1]       |■■■
0.057 [15]      |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
0.105 [10]      |■■■■■■■■■■■■■■■■■■■■■■■■■
0.154 [6]       |■■■■■■■■■■■■■■■
0.202 [10]      |■■■■■■■■■■■■■■■■■■■■■■■■■
0.251 [8]       |■■■■■■■■■■■■■■■■■■■■
0.299 [10]      |■■■■■■■■■■■■■■■■■■■■■■■■■
0.348 [11]      |■■■■■■■■■■■■■■■■■■■■■■■■■■■■
0.396 [5]       |■■■■■■■■■■■■■
0.445 [16]      |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
0.493 [8]       |■■■■■■■■■■■■■■■■■■■■

Latency distribution: 
  10% in 0.0344 secs 
  25% in 0.0984 secs 
  50% in 0.2525 secs 
  75% in 0.3884 secs 
  90% in 0.4407 secs 
  95% in 0.4698 secs 
  99% in 0.4934 secs 

Status code distribution: 
  [200] 96 responses
  [400] 1 responses
  [409] 1 responses
  [500] 1 responses
  [502] 1 responses
```
## Roadmap
The purpose of this project is primarily for learning, and I won't be investing more energy into it at present. However, there is a possibility that the following features may be developed in the future:
+ Improve result formatting.
+ Support more detailed statistical data, may include DNS resolution times, connection times, response times, etc.
+ Refactor using a better concurrency pattern.

CHANGELOG.md
# 0.1.0

+ Initial release

src/main.rs
use std::sync::Arc;
use std::time::{Duration, Instant};

use anyhow::{anyhow, Context};
use clap::Parser;
use flexi_logger::{FlexiLoggerError, Logger};
use reqwest::Url;
use tokio::signal::ctrl_c;
use tokio::sync::Notify;

use rey::arg::Args;
use rey::client::ClientBuilder;
use rey::work::Work;

macro_rules! unwrap_or_exit {
	($expr:expr) => {
		match $expr {
			Ok(val) => val,
			Err(err) => {
				eprintln!("{:?}", err);
				std::process::exit(1);
			}
		}
	};
}

#[tokio::main]
async fn main() {
	unwrap_or_exit!(init_logger().context("fail to statup logger"));
	let args = Args::parse();
	let body: Vec<u8>;
	if let Some(body_str) = args.body {
		body = body_str.into_bytes();
	} else if let Some(file) = args.body_file {
		body = unwrap_or_exit!(tokio::fs::read(file)
			.await
			.map_err(|err| { anyhow!("invalid BODY FILE: {}", err) }));
	} else {
		body = vec![];
	}
	let body: &'static [u8] = Box::leak(body.into_boxed_slice());
	let client_builder = ClientBuilder {
		headers: args.headers,
		timeout: if args.timeout > 0 {
			Some(Duration::from_secs(args.timeout))
		} else {
			None
		},
		content_type: args.content_type_header,
		accept: args.accept_header,
		user_agent: args.user_agent_header,
		proxy: args.proxy_address,
		host: args.host,
		disable_redirect: args.disable_redirect,
	};
	let work = Work {
		client_builder,
		url: unwrap_or_exit!(args.url.parse::<Url>().context("invalid url")),
		method: args.method,
		workers: args.workers,
		auth: args.basic_auth,
		total_requests: args.requests,
		rate_limit: args.rate_limit,
		body,
	};
	let notify = Arc::new(Notify::new());
	let cancel = notify.clone();
	// todo: instead by pending()
	let max_duration = args
		.max_duration
		.unwrap_or(Duration::from_secs(60 * 60 * 24));
	tokio::spawn(async move {
		tokio::select! {
			_ = tokio::time::sleep(max_duration)=>{}
			_ = ctrl_c()=>{}
		}
		notify.notify_one();
	});

	// execute
	let start = Instant::now();
	let report = unwrap_or_exit!(work.execute(cancel).await);
	let reporter = report.into_report(start.elapsed());
	reporter.print();
}

fn init_logger() -> Result<(), FlexiLoggerError> {
	Logger::try_with_env_or_str("warn")?
		.format(flexi_logger::colored_detailed_format)
		.start()?;
	Ok(())
}


src/work.rs
use std::collections::HashMap;
use std::sync::Arc;
use std::time::Duration;

use http::Method;
use log::info;
use reqwest::{Body, Client, Url};
use tokio::sync::mpsc::{channel, Sender};
use tokio::sync::Notify;
use tokio::time::Instant;

use crate::report::Reporter;

#[derive(Debug)]
struct SourceStat {
	pub duration: Duration,
	pub status_code: u16,
	pub content_length: u64,
}

type RequestResult = Result<SourceStat, reqwest::Error>;

#[derive(Debug, Clone, PartialEq)]
pub struct BasicAuth {
	pub username: String,
	pub password: Option<String>,
}

struct Worker<B>
where
	B: Into<Body> + Copy,
{
	url: Url,
	method: Method,
	basic_auth: Option<BasicAuth>,
	rate_limit: Option<f64>,
	body: B,
	requests: u64,
	client: Arc<Client>,
	sender: Sender<RequestResult>,
}

impl<B> Worker<B>
where
	B: Into<Body> + Copy,
{
	async fn make_request(&self) -> RequestResult {
		let start = Instant::now();

		// build
		let client = self.client.clone();
		let method = self.method.clone();
		let url = self.url.clone();
		let mut builder = client.request(method, url);
		if let Some(auth) = self.basic_auth.clone() {
			builder = builder.basic_auth(auth.username, auth.password);
		}
		// request
		let request = builder.body(self.body).build()?;
		let response = client.execute(request).await?;
		let status_code = response.status().as_u16();
		let content_length = response.content_length().unwrap_or(0);
		let _res = response.bytes().await?;
		Ok(SourceStat {
			duration: start.elapsed(),
			status_code,
			content_length,
		})
	}

	async fn execute(&self) {
		let interval = self
			.rate_limit
			.map(|qps| (1000000_f64 / qps).floor() as u64);
		for _ in 0..self.requests {
			if let Some(interval) = interval {
				tokio::time::sleep(Duration::from_micros(interval)).await;
			}
			let result = self.make_request().await;
			let sender = self.sender.clone();
			if let Err(error) = sender.send(result).await {
				info!("worker interrupt due to error:{}", error);
				return;
			}
		}
	}
}

pub struct Work<C, B>
where
	C: TryInto<Client, Error = anyhow::Error>,
	B: Into<Body> + Copy + Send + Sync + 'static,
{
	pub client_builder: C,
	pub url: Url,
	pub method: Method,
	pub auth: Option<BasicAuth>,
	pub workers: u16,
	pub total_requests: u64,
	pub rate_limit: Option<f64>,
	pub body: B,
}

impl<C, B> Work<C, B>
where
	C: TryInto<Client, Error = anyhow::Error>,
	B: Into<Body> + Copy + Send + Sync + 'static,
{
	pub async fn execute(self, cancel: Arc<Notify>) -> anyhow::Result<Reporter> {
		let client = Arc::new(self.client_builder.try_into()?);
		let requests = self.total_requests / (self.workers as u64);
		let (sender, mut receiver) = channel(self.workers as usize);
		for _ in 0..self.workers {
			let worker = Worker {
				url: self.url.clone(),
				method: self.method.clone(),
				basic_auth: self.auth.clone(),
				rate_limit: self.rate_limit,
				requests,
				client: client.clone(),
				sender: sender.clone(),
				body: self.body,
			};
			tokio::spawn(async move {
				worker.execute().await;
			});
		}
		drop(sender);

		let mut total_requests = 0_u64;
		let mut success_requests = 0_u64;
		let mut durations = vec![];
		let mut status_codes = vec![];
		let mut size_total = 0_u64;
		let mut error_dist = HashMap::new();

		loop {
			tokio::select! {
				_ = cancel.notified()=>{
					info!("receive cancel signal");
					receiver.close();
					break;
				}
				msg = receiver.recv() =>{
					match msg{
						None => {
							info!("all sender of worker been closed, finish receiving source stats");
							break;
						},
						Some(result)=>{
							total_requests += 1;
							match result{
								Err(err)=>*error_dist.entry(err.to_string()).or_insert(0) += 1,
								Ok(stat)=>{
									success_requests += 1;
									durations.push(stat.duration.as_secs_f64());
									status_codes.push(stat.status_code);
									size_total += stat.content_length;
								}
							}
						}
					}
				}
			}
		}

		Ok(Reporter {
			total_requests,
			success_requests,
			durations,
			status_codes,
			size_total,
			error_dist,
		})
	}
}


src/client.rs
use std::time::Duration;

use anyhow::{anyhow, Context, Error, Result};
use http::header::{ACCEPT, CONTENT_TYPE, HOST, USER_AGENT};
use http::{HeaderMap, HeaderName, HeaderValue};
use reqwest::redirect::Policy;
use reqwest::{Client, Proxy};

pub struct ClientBuilder {
	pub headers: Vec<String>,
	pub timeout: Option<Duration>,
	pub content_type: HeaderValue,
	pub accept: Option<HeaderValue>,
	pub user_agent: HeaderValue,
	pub proxy: Option<String>,
	pub host: Option<HeaderValue>,
	pub disable_redirect: bool,
}

impl TryFrom<ClientBuilder> for Client {
	type Error = anyhow::Error;

	fn try_from(cb: ClientBuilder) -> anyhow::Result<Client> {
		let mut headers = try_into_headers(&cb.headers)?;
		headers.insert(CONTENT_TYPE, cb.content_type);
		headers.insert(USER_AGENT, cb.user_agent);
		if let Some(accept) = cb.accept {
			headers.insert(ACCEPT, accept);
		}
		if let Some(host) = cb.host {
			headers.insert(HOST, host);
		}

		let mut builder = Client::builder();
		builder = builder.default_headers(headers);

		if let Some(timeout) = cb.timeout {
			builder = builder.timeout(timeout).connect_timeout(timeout);
		}
		if cb.disable_redirect {
			builder = builder.redirect(Policy::none())
		}
		if let Some(proxy) = cb.proxy {
			builder = builder.proxy(Proxy::all(proxy).context("invalid proxy")?);
		}
		builder.build().context("fail to build a http client")
	}
}

type Header = (HeaderName, HeaderValue);

fn try_into_headers(strs: &[String]) -> Result<HeaderMap, Error> {
	strs.iter()
		.map(|s| try_into_header(s))
		.collect::<Result<Vec<Header>, Error>>()
		.map(|headers| {
			headers
				.into_iter()
				.fold(HeaderMap::new(), |mut map, (name, value)| {
					map.insert(name, value);
					map
				})
		})
}

fn try_into_header(s: &str) -> Result<Header> {
	let parts: Vec<&str> = s.splitn(2, ':').collect();
	let get_error = || anyhow!("{} is not a valid header", s);
	if parts.len() != 2 {
		return Err(anyhow!("{} is not a valid header", s));
	}
	let name = HeaderName::try_from(parts[0].trim()).map_err(|_| get_error())?;
	let value = HeaderValue::try_from(parts[1].trim()).map_err(|_| get_error())?;
	Ok((name, value))
}

#[cfg(test)]
mod test {
	use crate::client::{try_into_header, try_into_headers};

	#[test]
	fn try_into_header_should_work() {
		let (name, value) = try_into_header("Token: abcdefg").unwrap();
		assert_eq!(name.as_str(), "token");
		assert_eq!(value.to_str().unwrap(), "abcdefg");
		let (name, value) = try_into_header("Accept-Language: gzip, deflate").unwrap();
		assert_eq!(name.as_str(), "accept-language");
		assert_eq!(value.to_str().unwrap(), "gzip, deflate");
	}

	#[test]
	fn try_into_headers_should_work() {
		let vec = vec![
			"Token: abcdefg".to_string(),
			"Accept-Language: gzip, deflate".to_string(),
		];
		let headers = try_into_headers(&vec).unwrap();
		assert_eq!(headers.get("Token").unwrap().to_str().unwrap(), "abcdefg");
		assert_eq!(
			headers.get("Accept-Language").unwrap().to_str().unwrap(),
			"gzip, deflate"
		);
	}
}


src/report.rs
use std::collections::HashMap;
use std::time::Duration;

use serde::{Deserialize, Serialize};
use tera::{to_value, try_get_value, Context, Filter, Tera, Value};

const BAR_CHAR: &str = "■";

const TEMPLATE: &str = r#"
Summary:
  Total:  {{ s.total | duration_to_sec_f64 | round(precision=4) }} secs
  Slowest:  {{ s.slowest | round(precision=4) }} secs
  Fastest:  {{ s.fastest | round(precision=4) }} secs
  Average:  {{ s.average | round(precision=4) }} secs
  Requests/sec:  {{ s.rps | round(precision=4) }}
  {% if s.size_total > 0 %}
  Total data:	{{ s.size_total | human_bytes }} bytes
  Size/request:	{{ s.size_req | human_bytes }} bytes {% endif %}

Response time histogram:
{{ s.histogram | histogram }}
Latency distribution: {% for dist in s.latency_dist %}
  {{ dist.percentage }}% in {{ dist.latency | round(precision=4) }} secs {% endfor %}

Status code distribution: {% for code, count in s.status_code_dist %}
  [{{ code }}]	{{ count }} responses{% endfor %}
{% if s.error_dist | length > 0 %}
Error distribution: {% for err, count in s.error_dist %}
  [{{ count }}] {{ err }}{% endfor %}{% endif %}
"#;

#[derive(Debug, Default, Serialize)]
pub struct LatencyDistribution {
	percentage: u8,
	latency: f64,
}

#[derive(Debug, Default, Serialize, Deserialize)]
pub struct Bucket {
	pub mark: f64,
	pub count: u64,
	pub frequency: f64,
}

struct DurationToSecF64Filter;

impl Filter for DurationToSecF64Filter {
	fn filter(&self, value: &Value, _args: &HashMap<String, Value>) -> tera::Result<Value> {
		let duration = try_get_value!("duration_to_sec_f64", "value", Duration, value);
		Ok(to_value(duration.as_secs_f64())?)
	}
}

struct HumanBytesFilter;

impl Filter for HumanBytesFilter {
	fn filter(&self, value: &Value, _args: &HashMap<String, Value>) -> tera::Result<Value> {
		let bytes = try_get_value!("human_bytes", "value", u64, value);
		Ok(to_value(human_bytes::human_bytes(bytes as f64))?)
	}
}

struct HistogramFilter;

impl Filter for HistogramFilter {
	fn filter(&self, value: &Value, _args: &HashMap<String, Value>) -> tera::Result<Value> {
		let buckets = try_get_value!("histogram", "value", Vec<Bucket>, value);
		let max = buckets.iter().map(|bucket| bucket.count).max();
		let mut string = String::default();
		for ref bucket in buckets {
			let bar = max
				.map(|value| (bucket.count * 40 + value / 2) / value)
				.map(|len| BAR_CHAR.repeat(len as usize))
				.unwrap_or_default();
			string
				.push_str(format!("{:>4.3} [{}]\t|{}\n", bucket.mark, bucket.count, bar).as_str());
		}
		Ok(to_value(string)?)
	}
}

#[derive(Debug, Default, Serialize)]
pub struct Report {
	pub avg_total: f64,
	pub fastest: f64,
	pub slowest: f64,
	pub average: f64,
	pub rps: f64,

	pub total_requests: u64,

	pub total: Duration,

	pub error_dist: HashMap<String, u64>,
	pub status_code_dist: HashMap<u16, u64>,
	pub size_total: u64,
	pub size_req: u64,
	pub num_res: u64,

	pub latency_dist: Vec<LatencyDistribution>,
	pub histogram: Vec<Bucket>,
}

impl Report {
	pub fn print(&self) {
		let mut ctx = Context::new();
		ctx.insert("s", self);
		let mut tera = Tera::default();
		tera.register_filter("duration_to_sec_f64", DurationToSecF64Filter);
		tera.register_filter("human_bytes", HumanBytesFilter);
		tera.register_filter("histogram", HistogramFilter);
		let string = tera.render_str(TEMPLATE, &ctx).unwrap();
		println!("{}", string);
	}
}

#[derive(Default)]
pub struct Reporter {
	pub total_requests: u64,
	pub success_requests: u64,
	pub status_codes: Vec<u16>,
	pub size_total: u64,
	pub error_dist: HashMap<String, u64>,
	pub durations: Vec<f64>,
}

impl Reporter {
	fn histogram(&self, fastest: f64, slowest: f64) -> Vec<Bucket> {
		if self.success_requests == 0 {
			return vec![];
		}
		let bc = 10_usize;
		let mut buckets: Vec<f64> = Vec::with_capacity(bc + 1);
		let mut counts: Vec<u64> = vec![0; bc + 1];
		let bs = (slowest - fastest) / bc as f64;
		for i in 0..bc {
			buckets.push(fastest + bs * (i as f64));
		}
		buckets.push(slowest);

		let mut bi = 0_usize;
		let mut max = 0_u64;
		let mut i = 0_usize;
		while i < self.durations.len() {
			if self.durations[i] <= buckets[bi] {
				i += 1;
				counts[bi] += 1;
				if max < counts[bi] {
					max = counts[bi]
				}
			} else if bi < buckets.len() - 1 {
				bi += 1;
			}
		}

		buckets
			.iter()
			.zip(counts.iter())
			.map(|(bucket, count)| Bucket {
				mark: *bucket,
				count: *count,
				frequency: (*count) as f64 / self.durations.len() as f64,
			})
			.collect()
	}

	fn latencies(&self) -> Vec<LatencyDistribution> {
		let pctls = [10_u8, 25, 50, 75, 90, 95, 99];
		let mut data = vec![];
		let mut i = 0_usize;
		let mut j = 0_usize;
		while i < self.durations.len() && data.len() < pctls.len() {
			let current = i * 100 / self.durations.len();
			if current >= pctls[j] as usize {
				data.push(self.durations[i]);
				j += 1;
			}
			i += 1;
		}
		pctls
			.iter()
			.zip(data.iter())
			.map(|(p, d)| LatencyDistribution {
				percentage: *p,
				latency: *d,
			})
			.collect()
	}

	pub fn into_report(mut self, total: Duration) -> Report {
		let mut report = Report {
			total,
			rps: self.total_requests as f64 / total.as_secs_f64(),
			avg_total: self.durations.iter().sum(),
			total_requests: self.total_requests,
			size_total: self.size_total,
			..Report::default()
		};
		if self.success_requests > 0 {
			report.average = report.avg_total / self.success_requests as f64;
			report.size_req = self.size_total / self.success_requests;
		} else {
			report.average = 0.0;
			report.size_req = 0;
		}

		self.durations.sort_by(|a, b| a.total_cmp(b));
		report.fastest = *self.durations.first().unwrap_or(&0.0);
		report.slowest = *self.durations.last().unwrap_or(&0.0);
		report.histogram = self.histogram(report.fastest, report.slowest);
		report.latency_dist = self.latencies();
		report.error_dist = self.error_dist;
		report.status_code_dist =
			self.status_codes
				.into_iter()
				.fold(HashMap::new(), |mut map, code| {
					*map.entry(code).or_insert(0) += 1;
					map
				});

		report
	}
}


src/arg.rs
use std::collections::HashSet;
use std::path::PathBuf;
use std::time::Duration;

use anyhow::Result;
use clap::Parser;
use http::{HeaderValue, Method};
use lazy_static::lazy_static;

use crate::work::BasicAuth;

lazy_static! {
	static ref VALID_METHODS: HashSet<Method> = {
		let mut set = HashSet::new();
		set.insert(Method::GET);
		set.insert(Method::POST);
		set.insert(Method::PUT);
		set.insert(Method::DELETE);
		set.insert(Method::HEAD);
		set.insert(Method::PATCH);
		set.insert(Method::OPTIONS);
		set.insert(Method::TRACE);
		set.insert(Method::CONNECT);
		set
	};
}

fn parse_duration(s: &str) -> Result<Duration, &'static str> {
	duration_str::parse(s).map_err(|_err| "not a standard duration string")
}

fn parse_method(s: &str) -> Result<Method, &'static str> {
	match s.parse::<Method>() {
		Err(_) => Err("invalid method"),
		Ok(ref method) if !VALID_METHODS.contains(method) => Err("invalid method"),
		Ok(method) => Ok(method),
	}
}

fn parse_basic_auth(s: &str) -> Result<BasicAuth, &'static str> {
	let splits: Vec<&str> = s.splitn(2, ':').collect();
	if splits.is_empty() {
		return Err("invalid username and password");
	}
	Ok(BasicAuth {
		username: splits[0].to_string(),
		password: splits.get(1).map(|password| password.to_string()),
	})
}

macro_rules! define_parse_header_fn {
	($fn_name:ident, $static_str: expr) => {
		fn $fn_name(s: &str) -> Result<HeaderValue, &'static str> {
			HeaderValue::try_from(s).map_err(|_| $static_str)
		}
	};
}

define_parse_header_fn!(parse_content_type, "invalid content-type");
define_parse_header_fn!(parse_accept, "invalid accept");
define_parse_header_fn!(parse_user_agent, "invalid user agent");
// define_parse_header_fn!(parse_host, "invalid host");

#[derive(Parser, Debug)]
#[command(version)]
pub struct Args {
	pub url: String,

	/// Name of the person to greet
	#[arg(short = 'n', default_value = "200")]
	pub requests: u64,

	/// Number of workers to run concurrently. Total number of requests cannot be smaller than the concurrency level
	#[arg(short = 'c', default_value = "50")]
	pub workers: u16,

	/// Rate limit, in queries per second (QPS) per worker
	#[arg(short = 'q', value_name = "RATE LIMIT")]
	pub rate_limit: Option<f64>,

	/// Duration of application to send requests. When duration is reached, application stops and exits. If duration is specified, n is ignored. Examples: -z 10s -z 3m
	#[arg(short = 'z', value_name = "Duration", value_parser = parse_duration)]
	pub max_duration: Option<Duration>,

	/// HTTP method, one of GET, POST, PUT, DELETE, HEAD, OPTIONS
	#[arg(short = 'm', value_parser = parse_method, default_value = "GET")]
	pub method: Method,

	/// Custom HTTP header. You can specify as many as needed by repeating the flag. For example, -H "Accept: text/html" -H "Content-Type: application/xml"
	#[arg(short = 'H', action = clap::ArgAction::Append)]
	pub headers: Vec<String>,

	/// Timeout for each request in seconds. Use 0 for infinite
	#[arg(short = 't', default_value = "20")]
	pub timeout: u64,

	/// HTTP Accept header
	#[arg(short = 'A', value_name = "ACCEPT HEADER", value_parser = parse_accept)]
	pub accept_header: Option<HeaderValue>,

	/// Content-type, defaults to "text/html"
	#[arg(short = 'T', value_name = "CONTENT-TYPE", default_value = "text/html", value_parser = parse_content_type)]
	pub content_type_header: HeaderValue,

	/// User-Agent, defaults to version "rey/0.1.0"
	#[arg(short = 'U', value_name = "USER AGENT", default_value = "rey/0.1.0", value_parser = parse_user_agent)]
	pub user_agent_header: HeaderValue,

	/// HTTP request body
	#[arg(short = 'd')]
	pub body: Option<String>,

	/// HTTP request body from file. For example, /home/user/file.txt or ./file.txt
	#[arg(short = 'D', value_name = "FILE")]
	pub body_file: Option<PathBuf>,

	/// Basic authentication, username:password
	#[arg(short = 'a', value_name = "USERNAME:PASSWORD", value_parser = parse_basic_auth)]
	pub basic_auth: Option<BasicAuth>,

	/// HTTP Proxy address as host:port
	#[arg(short = 'x', value_name = "PROXY")]
	pub proxy_address: Option<String>,

	#[arg(long = "host", value_name = "HOST")]
	pub host: Option<HeaderValue>,

	#[arg(
		long = "disable-redirects",
		value_name = "DISABLE REDIRECT",
		default_value = "false"
	)]
	pub disable_redirect: bool,
}

#[cfg(test)]
mod tests {
	use http::Method;

	use crate::arg::{
		parse_accept, parse_basic_auth, parse_content_type, parse_method, parse_user_agent,
	};
	use crate::work::BasicAuth;

	#[test]
	fn parse_method_should_work() {
		assert_eq!(Ok(Method::GET), parse_method("GET"));
		assert_eq!(Ok(Method::POST), parse_method("POST"));
		assert_eq!(Ok(Method::PUT), parse_method("PUT"));
		assert_eq!(Ok(Method::DELETE), parse_method("DELETE"));
		assert_eq!(Ok(Method::HEAD), parse_method("HEAD"));
		assert_eq!(Ok(Method::OPTIONS), parse_method("OPTIONS"));
		assert_eq!(Ok(Method::CONNECT), parse_method("CONNECT"));
		assert_eq!(Ok(Method::TRACE), parse_method("TRACE"));
		assert_eq!(Ok(Method::PATCH), parse_method("PATCH"));
	}

	#[test]
	fn parse_method_return_error() {
		assert_eq!(Err("invalid method"), parse_method(""));
		assert_eq!(Err("invalid method"), parse_method("HELLO"));
		assert_eq!(Err("invalid method"), parse_method("get"));
		assert_eq!(Err("invalid method"), parse_method("Get"));
		assert_eq!(Err("invalid method"), parse_method("大便"));
	}

	#[test]
	fn parse_header_should_work() {
		assert_eq!(parse_accept("*/*").unwrap().to_str().unwrap(), "*/*");
		assert_eq!(
			parse_content_type("application/json")
				.unwrap()
				.to_str()
				.unwrap(),
			"application/json"
		);
		assert_eq!(parse_user_agent("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3").unwrap(),
                   "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3");
	}

	#[test]
	fn parse_basic_auth_should_work() {
		assert_eq!(
			parse_basic_auth("root:123456"),
			Ok(BasicAuth {
				username: "root".to_string(),
				password: Some("123456".to_string())
			})
		);
		assert_eq!(
			parse_basic_auth("root"),
			Ok(BasicAuth {
				username: "root".to_string(),
				password: None
			})
		);
	}
}
